### ISSUE TITLE
Don't use `InitializeParams.rootUri` if `workspaceFolders` is specified

### ISSUE THREAD

**Issue by alexander-doroshko** [tailwindlabs/tailwindcss-intellisense#950]
Hi from JetBrains,

The LSP spec deprecates [InitializeParams.rootUri](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#initializeParams) in favour of `workspaceFolders`.

I noticed that `rootUri` is used unconditionally in the [tw.ts](https://github.com/tailwindlabs/tailwindcss-intellisense/blob/daa8bb2821af9cffd4d3a0e2c5982aeea0e2bd7e/packages/tailwindcss-language-server/src/tw.ts#L114) file.

This causes the following problem. The IDE sends the `initialize` request as follows:
```
{
  "method": "initialize",
  "params": {
    ...
    "rootUri": null,
    "workspaceFolders": [
      {
        "uri": "file:///foo/bar",
        "name": "bar"
      },
      {
        "uri": "file:///foo/baz",
        "name": "baz"
      }
    ]
  }
}
```

and the server fails to initialize, it says `No workspace folders found, not initializing.`

In my understanding, if `workspaceFolders` is given, the server should ignore `rootUri`.

**Comment by thecrypticace** [comment]
This one was already on my radar but thanks for the reminder! We don't look at `workspaceFolders` at all right now and we definitely should. I noticed the deprecations after updating the LSP-related packages. There's some things I need to think through I think about how things need to work when using them but I'm planning to do so this week if I can — otherwise next week.

**Comment by alexander-doroshko** [comment]
Thanks. I've worked it around in the JetBrains Tailwind CSS plugin by starting a separate LSP server for each root. But I'll be happy to remove this workaround on our end once the LSP server supports `workspaceFolders`.

**Comment by thecrypticace** [comment]
@alexander-doroshko Sounds good!

Feel free to ping me if you run into other stuff or have questions/ideas/suggestions 👍

**Comment by thecrypticace** [comment]
Can you test the latest insiders build of `@tailwindcss/language-server` to see if things work as you expect?

**Comment by alexander-doroshko** [comment]
Cool, [0.0.0-insiders.830dc0a](https://www.npmjs.com/package/@tailwindcss/language-server/v/0.0.0-insiders.830dc0a) works for me. I tried a simple scenario with a project with two roots. I haven't checked how it handles multiple `package.json` files, multiple `node_modules` folders, different `tailwindcss` package versions in different workspace roots, or multiple config files. I trust it works perfectly as usual :)

**Comment by thecrypticace** [comment]
Hmm it definitely should. I'll take a look.

Any chance you could provide the logs and/or messages are you getting from the server?

**Comment by thecrypticace** [comment]
Weird — that's helpful! I'll see if I can figure out what's wrong today.

**Comment by thecrypticace** [comment]
Oh wait it could be the presence of the tailwind config file. That's not supported with v4 right now.

**Comment by thecrypticace** [comment]
Oh no it's actually that you don't have any css files that import tailwind in some manner. With v4 you need a CSS file in your project that contains `@import "tailwindcss"`. We analyze all the CSS files in the project to look for multiple configs and follow the import graph for theme declarations.

**Comment by alexander-doroshko** [comment]
I've caught this once. I think I was moving an HTML file from one folder to another when it happened. Thought I couldn't reproduce it again.

```
Tailwind CSS: Unhandled rejection TypeError: Cannot read properties of null (reading 'match')
at Klt (.../tailwindcss-language-server:2920:15716)
at async d (.../tailwindcss-language-server:2920:18890)
at async .../tailwindcss-language-server:2920:19607
```

**Comment by thecrypticace** [comment]
Yeah, right now if the LSP is already running and you add the CSS file it won't get picked up. Definitely some improvements to be made there.

> I've caught this once. I think I was moving an HTML file from one folder to another when it happened. Thought I couldn't reproduce it again.

A read of the code suggests that it was the CSS file that moved. Fix is here #956 :)

This bug is a good showcase for why I'm working to (slowly) migrate this codebase to strict types.

**Comment by alexander-doroshko** [comment]
> A read of the code suggests that it was the CSS file that moved

I think I didn't have any CSS files in the project at that moment. Only those 3: package.json, tailwind.config.js, and index.html. And node_modules folder. IAFAIR, I moved HTML file from the project root folder to src.

I guess #956 fixes that NPE anyway.
